### PR TITLE
[core]Trim `__FILE__` macro by compiler if supported

### PR DESCRIPTION
**Pull request by praasz**
### Details:
 - For compilers which support the trimming of `__FILE__` macro the file names are reduced to have got relative path to project root folder.
 - Compiler which not supported trim `__FILE__` macro keep original `__FILE__` value but display e.g. in exception to pointer to file name path is shifted by size of root project path name.
 
 For build version which support `__FILE__` trim the binary size of libraries is reduce e.g. (Ubuntu 20.04, gcc 9.3 Release version no LTO) is about:
 * openvino reduced by 14.55 KiB
 * intel_cpu reduce by 8KiB

### Tickets:
 - 113448

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -5,6 +5,25 @@
...
+add_compile_options(
+
+    # Enable __FILE__ trimming by compiler for:
+    # AppleClang, Clang, GNU
+    $<$<COMPILE_LANG_AND_ID:C,AppleClang,Clang,GNU>:-ffile-prefix-map=${NATIVE_PROJECT_ROOT_DIR}/=>
+    $<$<COMPILE_LANG_AND_ID:CXX,AppleClang,Clang,GNU>:-ffile-prefix-map=${NATIVE_PROJECT_ROOT_DIR}/=>
+
+    # MSVC
+    $<$<COMPILE_LANG_AND_ID:C,MSVC>:/d1trimfile:${NATIVE_PROJECT_ROOT_DIR}\\>
+    $<$<COMPILE_LANG_AND_ID:CXX,MSVC>:/d1trimfile:${NATIVE_PROJECT_ROOT_DIR}\\>)
```
Comment (ilya-lavrenov) [inline src/CMakeLists.txt:25]: should we move this code into cmake/developer_package/os_flags.cmake ?
In this case plugin which are built on top of prebuilt OpenVINO will also benefit from this change.

File: src/core/include/ngraph/log.hpp [hunk 2]
```diff
@@ -70,7 +70,11 @@ constexpr const char* get_file_name(ConstString s) {
 }
 NGRAPH_API_DEPRECATED
 constexpr const char* trim_file_name(ConstString root, ConstString s) {
+#ifdef PROJECT_ROOT_DIR_LENGTH
```
Comment (ilya-lavrenov) [inline src/core/include/ngraph/log.hpp:73]: it's a public header.
when samples are compiled against prebuilt OpenVINO, this macro is not defined. So, should we move implementation into .cpp ?

File: src/CMakeLists.txt (hunk 1 above)
Comment (praasz) [reply src/CMakeLists.txt:25]: Yes, sure. I was not sure if this folder is correct.

**Comment by praasz** [comment]
@ilya-lavrenov 
Is there any list what minimum version of cmake, compilers should be supported?
 I need find some other solution for version where some new features are not supported for older versions.

**Comment by ilya-lavrenov** [comment]
> @ilya-lavrenov
> Is there any list what minimum version of cmake, compilers should be supported?
> I need find some other solution for version where some new features are not supported for older versions.

We support:
- MSVC 2019 and newer
- GCC 7 and newer
- Clang and AppleClang - not sure about exact versions.

File: src/common/util/include/openvino/util/log.hpp [hunk 3]
```diff
@@ -85,36 +83,32 @@ class Logger {
 
 void default_logger_handler_func(const std::string& s);
 
-#ifndef PROJECT_ROOT_DIR
```
Comment (ilyachur) [inline src/common/util/include/openvino/util/log.hpp:88]: Should we remove: https://github.com/openvinotoolkit/openvino/blob/master/src/CMakeLists.txt#L6 ?

Comment (praasz) [reply src/common/util/include/openvino/util/log.hpp:88]: Removed. Is not required any more.

File: src/bindings/c/include/c_api/ie_c_api.h [hunk 4]
```diff
@@ -73,1092 +73,1128 @@
...
+             * @param core A pointer to the ie_core_t instance.
+             * @param file_name A path to the location of the exported file.
+             * @param device_name A name of the device to load the network to.
+             * @param config Device configuration.
+             * @param exe_network A pointer to the newly created executable network.
+             * @return Status code of the operation: OK(0) for success.
+             */
+            INFERENCE_ENGINE_C_API(IE_NODISCARD IEStatusCode)
+            ie_core_import_network(ie_core_t* core,
+                                   const char* file_name,
```
Comment (ilya-lavrenov) [inline src/bindings/c/include/c_api/ie_c_api.h:445]: I suppose we need to revert changes in this file?

Comment (praasz) [reply src/bindings/c/include/c_api/ie_c_api.h]: Yes, reverted.

File: cmake/developer_package/compile_flags/os_flags.cmake [hunk 5]
```diff
@@ -318,6 +318,18 @@ elseif(CMAKE_COMPILER_IS_GNUCXX OR OV_COMPILER_IS_CLANG)
     ie_add_compiler_flags(-fsigned-char)
 endif()
 
+file(RELATIVE_PATH OV_RELATIVE_BIN_PATH ${CMAKE_CURRENT_BINARY_DIR} ${CMAKE_SOURCE_DIR})
```
Comment (ilya-lavrenov) [inline cmake/developer_package/compile_flags/os_flags.cmake:321]: does it imply that binary folder is located inside the source dir?  If no, what is logic here?

File: cmake/developer_package/compile_flags/os_flags.cmake [hunk 6]
```diff
@@ -430,9 +445,22 @@ else()
     # - https://gcc.gnu.org/onlinedocs/gcc/C_002b_002b-Dialect-Options.html
     # - https://gcc.gnu.org/onlinedocs/libstdc++/manual/abi.html
     if(CMAKE_COMPILER_IS_GNUCXX)
+        if(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL "8")
+            # Enable __FILE__ trim
+            ie_add_compiler_flags(-ffile-prefix-map=${OV_NATIVE_PROJECT_ROOT_DIR}/=)
+            ie_add_compiler_flags(-ffile-prefix-map=${OV_RELATIVE_BIN_PATH}/=)
```
Comment (ilya-lavrenov) [inline cmake/developer_package/compile_flags/os_flags.cmake:451]: do we also need to use -fmacro-prefix-map or -fdebug-prefix-map ?

Comment (praasz) [reply cmake/developer_package/compile_flags/os_flags.cmake:451]: specify this option will specify all `-f*-prefix-map`

**Review by ilya-lavrenov — APPROVED** [review]

**Pull request merged**